src/ir.cpp

```diff
@@ -20653,23 +20653,33 @@ static IrInstGen *ir_analyze_instruction_elem_ptr(IrAnalyze *ira, IrInstSrcElemP
     // We will adjust return_type's alignment before returning it.
     ZigType *return_type;
 
-    if (type_is_invalid(array_type)) {
+    if (type_is_invalid(array_type))
         return ira->codegen->invalid_inst_gen;
-    } else if (array_type->id == ZigTypeIdArray ||
-        (array_type->id == ZigTypeIdPointer &&
-         array_type->data.pointer.ptr_len == PtrLenSingle &&
-         array_type->data.pointer.child_type->id == ZigTypeIdArray))
+
+    if (array_type->id == ZigTypeIdPointer &&
+        array_type->data.pointer.ptr_len == PtrLenSingle &&
+        array_type->data.pointer.child_type->id == ZigTypeIdArray)
     {
-        if (array_type->id == ZigTypeIdPointer) {
-            array_type = array_type->data.pointer.child_type;
-            ptr_type = ptr_type->data.pointer.child_type;
-            if (orig_array_ptr_val->special != ConstValSpecialRuntime) {
-                orig_array_ptr_val = const_ptr_pointee(ira, ira->codegen, orig_array_ptr_val,
-                        elem_ptr_instruction->base.base.source_node);
-                if (orig_array_ptr_val == nullptr)
-                    return ira->codegen->invalid_inst_gen;
-            }
+        array_type = array_type->data.pointer.child_type;
+        ptr_type = ptr_type->data.pointer.child_type;
+
+        if (orig_array_ptr_val->special != ConstValSpecialRuntime &&
+            orig_array_ptr_val->data.x_ptr.special != ConstPtrSpecialHardCodedAddr)
+        {
+            orig_array_ptr_val = const_ptr_pointee(ira, ira->codegen, orig_array_ptr_val,
+                elem_ptr_instruction->base.base.source_node);
+            if (orig_array_ptr_val == nullptr)
+                return ira->codegen->invalid_inst_gen;
         }
+    }
+
+    if (orig_array_ptr_val->special == ConstValSpecialUndef) {
```

Comment:
I think a call to `ir_resolve_const_val` is appropriate here. Maybe like this (changing some lines above as well):

```c++
if (orig_array_ptr_val->special != ConstValSpecialRuntime) {
    if ((err = ir_resolve_const_val(ira->codegen, ira->new_irb.exec, value->base.source_node,
                    orig_array_ptr_val, UndefBad)))
    {
        return ira->codegen->invalid_inst_gen;
    }
    if (orig_array_ptr_val->data.x_ptr.special != ConstPtrSpecialHardCodedAddr) {
        orig_array_ptr_val = const_ptr_pointee(ira, ira->codegen, orig_array_ptr_val,
            elem_ptr_instruction->base.base.source_node);
        if (orig_array_ptr_val == nullptr)
            return ira->codegen->invalid_inst_gen; 
        // and then if we need to make sure the new orig_array_ptr_val is also not
        // undefined then another call to ir_resolve_const_val here.
    }
}
```